src/bvh/bvh_node.rs

```diff
@@ -0,0 +1,608 @@
+
+use crate::ray::Ray;
+use crate::utils::{joint_aabb_of_shapes, Bucket};
+use std::cell::RefCell;
+use std::marker::PhantomData;
+use std::mem::MaybeUninit;
+
+const NUM_BUCKETS: usize = 6;
+
+thread_local! {
+    /// Thread local for the buckets used while building to reduce allocations during build
+    static BUCKETS: RefCell<[Vec<ShapeIndex>; NUM_BUCKETS]> = RefCell::new(Default::default());
+}
+
+/// The [`BvhNode`] enum that describes a node in a [`Bvh`].
+/// It's either a leaf node and references a shape (by holding its index)
+/// or a regular node that has two child nodes.
+/// The non-leaf node stores the [`Aabb`]s of its children.
+///
+/// [`Aabb`]: ../aabb/struct.Aabb.html
+/// [`Bvh`]: struct.Bvh.html
+/// [`Bvh`]: struct.BvhNode.html
+///
+#[derive(Debug, Copy, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum BvhNode<T: BHValue, const D: usize> {
+    /// Leaf node.
+    Leaf {
+        /// The node's parent.
+        parent_index: usize,
+
+        /// The shape contained in this leaf.
+        shape_index: usize,
+    },
+    /// Inner node.
+    Node {
+        /// The node's parent.
+        parent_index: usize,
+
+        /// Index of the left subtree's root node.
+        child_l_index: usize,
+
+        /// The convex hull of the shapes' [`Aabb`]'s in child_l.
+        child_l_aabb: Aabb<T, D>,
+
+        /// Index of the right subtree's root node.
+        child_r_index: usize,
+
+        /// The convex hull of the shapes' [`Aabb`]'s in child_r.
+        child_r_aabb: Aabb<T, D>,
+    },
+}
+
+impl<T: BHValue, const D: usize> BvhNode<T, D> {
+    /// Builds a [`BvhNode`] recursively using SAH partitioning.
+    ///
+    /// [`BvhNode`]: enum.BvhNode.html
+    ///
+    pub fn build<S: BHShape<T, D>>(args: BvhNodeBuildArgs<S, T, D>) {
+        if let Some((left, right)) = Self::prep_build(args) {
+            Self::build(left);
+            Self::build(right);
+            // BvhBuildStrategy::<X, Y>::dispatch(|| Self::build(left, strategy), || Self::build(right, strategy), left.indices.len() + right.indices.len(), left.depth as usize)
+        }
+    }
+
+    /// Builds a [`BvhNode`] recursively in parallel using SAH partitioning.
+    ///
+    /// [`BvhNode`]: enum.BvhNode.html
+    ///
+    pub fn build_with_executor<S: BHShape<T, D>>(
+        args: BvhNodeBuildArgs<S, T, D>,
+        mut executor: impl FnMut(BvhNodeBuildArgs<S, T, D>, BvhNodeBuildArgs<S, T, D>),
+    ) {
+        if let Some((left, right)) = Self::prep_build(args) {
+            // Self::build(left);
+            // Self::build(right);
+            executor(left, right);
+            // RayonBuildStrategy::dispatch(|| Self::build(left), || Self::build(right), left.indices.len() + right.indices.len(), left.depth as usize)
+        }
+    }
+
+    /// Builds a [`BvhNode`] recursively using SAH partitioning.
+    ///
+    /// [`BvhNode`]: enum.BvhNode.html
+    ///
```

Comment:
Could you document the meaning of the return value?

Reply:
Done